clustering_code/edchan-kmeans.py

```diff
@@ -0,0 +1,913 @@
+
+# %%
+def decode_qa(bit_string):
+    """
+    Decode an 8-bit string into its quality assessment components.
+
+    Bits are listed from the MSB (bit 7) to the LSB (bit 0): 
+        7-6    aerosol:
+            00 - climatology
+            01 - low
+            10 - average
+            11 - high
+        5      water
+        4      snow/ice
+        3      cloud shadow
+        2      adjacent to cloud
+        1      cloud
+        0      cirrus cloud
+    
+    Parameters:
+    bit_string (str): An 8-bit string representing quality assessment flags.
+    
+    Returns:
+    dict: A dictionary with human-readable descriptions of the QA flags.
+    """
+    # Define the bit positions based on the provided table
+    aerosol_level_bits = bit_string[:2]
+    water_bit = bit_string[2]
+    snow_ice_bit = bit_string[3]
+    cloud_shadow_bit = bit_string[4]
+    adjacent_cloud_shadow_bit = bit_string[5]
+    cloud_bit = bit_string[6]
+    
+    # Map the bit values to their meanings
+    aerosol_levels = {
+        '00': 'Climatology aerosol',
+        '01': 'Low aerosol',
+        '10': 'Moderate aerosol',
+        '11': 'High aerosol'
+    }
+    
+    # Decode the bits using the mapping
+    aerosol_level = aerosol_levels[aerosol_level_bits]
+    water = 'Yes' if water_bit == '1' else 'No'
+    snow_ice = 'Yes' if snow_ice_bit == '1' else 'No'
+    cloud_shadow = 'Yes' if cloud_shadow_bit == '1' else 'No'
+    adjacent_cloud_shadow = 'Yes' if adjacent_cloud_shadow_bit == '1' else 'No'
+    cloud = 'Yes' if cloud_bit == '1' else 'No'
+    
+    # Create a dictionary of the decoded values
+    decoded_values = {
+        'Aerosol Level': aerosol_level,
+        'Water': water,
+        'Snow/Ice': snow_ice,
+        'Cloud Shadow': cloud_shadow,
+        'Adjacent to Cloud/Shadow': adjacent_cloud_shadow,
+        'Cloud': cloud
+    }
+    
+    return decoded_values
+
+# %%
+water_bit = 2
+snow_ice_bit = 3
+cloud_shadow_bit = 4
+adjacent_cloud_shadow_bit = 5
+cloud_bit = 6
+mask_bits = [cloud_bit]
+temp_gdf = gpd.GeoDataFrame()
+processed_granules = []
+
+for date, gdf_by_date in filtered_gdf.groupby('date'):
+    print(f'Processing date: {date}')
+    fmask_das = []
+    red_das = []
+    green_das = []
+    blue_das = []
+    nir_das = []
+    swir1_das = []
+    swir2_das = []
+    merged_fmask = []
+    merged_red = []
+    merged_green = []
+    merged_blue = []
+    merged_nir = []
+    merged_swir1 = []
+    merged_swir2 = []
+    for band, gdf_by_band in gdf_by_date.groupby('band'):
+        # print(f'Processing band: {band}')
+        for index, row in gdf_by_band.iterrows():
+            filename = row.filename
+            folder = row.granule_id
+            filepath = os.path.join(cache_dir, folder, filename)
+            # print(filepath, row.band)
+            # Set masked=True to convert -9999 to NaN
+            if row.band == 'B02':
```

Comment:
I'd suggest looping here - this is repetitive and would be a lot of busy work to change if you use a different dataset with different bands. You can make a dictionary to rename the bands instead, and then this will be less repetitive. You could even use the conditionals for the file name if you want, but have the functions that get repeated outside the conditional.